build/generate-go-licenses.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+//go:build ignore
+
+package main
+
+import (
+	"encoding/json"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+)
+
+// regexp is based on go-license, excluding README and NOTICE
+// https://github.com/google/go-licenses/blob/master/licenses/find.go
+var licenseRe = regexp.MustCompile(`^(?i)((UN)?LICEN(S|C)E|COPYING).*$`)
+
+type LicenseEntry struct {
+	Name        string `json:"name"`
+	LicenseText string `json:"licenseText"`
+}
+
+func main() {
+	base, out := os.Args[1], os.Args[2]
+
+	paths := []string{}
+	err := filepath.WalkDir(base, func(path string, entry fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if entry.IsDir() || !licenseRe.MatchString(entry.Name()) {
+			return nil
+		}
+		paths = append(paths, path)
+		return nil
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	sort.Strings(paths)
+
+	entries := []LicenseEntry{}
+	for _, path := range paths {
+		licenseText, err := os.ReadFile(path)
+		if err != nil {
+			panic(err)
+		}
+
+		entries = append(entries, LicenseEntry{
+			Name:        filepath.Dir(strings.Replace(path, base+string(os.PathSeparator), "", 1)),
```

Comment:
I think it might be beneficial to post the full path of the license here as well.
This should even integrate nicely into the existing mechanism without requiring further changes.

Reply:
Yeah, that was also not intended.
But it helps if we ever want to see the license path for any reason, because the information is already there, and the data format allows us to add that without breaking, so why not use it?

Reply:
Ah, I see what you mean, well yeas, can just remove the `filepath.Dir` so the license filename shows in the output. I had been cutting it off as a go module name is usually idenfied by a URL path, excluding any filenames.

Reply:
Examples:

```diff
-    "name": "github.com/prometheus/client_golang/prometheus",
+    "name": "github.com/prometheus/client_golang/prometheus/LICENSE",
```
```
-    "name": "github.com/prometheus/common",
+    "name": "github.com/prometheus/common/LICENSE",
```
I'm actually not sure we want this as a go module is generally named by its import path which I was trying to replicate.

Reply:
I'll add it as a unused JSON property.

Reply:
> Ah, I see what you mean

Do you? Because now you confused me.

If I understand you correctly, then you can ignore my suggestion.

Reply:
Your comment's didn't show up for some reason. In any case, it's added now. I agree it might be helpful to see it to troubleshoot.

Reply:
> I'll add it as a unused JSON property.

Are we really talking about the same thing?
That was what I wanted the whole time… :laughing:

Reply:
Your comment was "post the full path" in the line with `Name`, so I took it you wanted it replaced :D

Reply:
No, no, I simply wanted an extra variable `Path: path`…
Sorry for the misunderstanding.